### PR TITLE
Change TargetAllocator configmap for ScrapeConfig and Probe CR

### PR DESCRIPTION
**Pull request by dominicqi**
**Description:**  Enable ScrapeConfigSelector and ProbeSelector support in OpenTelemetryTargetAllocatorPrometheusCR, create right configmap for target allocator

**Link to tracking Issue(s):** <Issue number if applicable>

- Resolves: #1842 

**Testing:** add test in `configmap_test.go`

**Documentation:** ScrapeConfigSelector and ProbeSelector config  in v1beta1 api doc

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -378,7 +378,7 @@ scorecard-tests: operator-sdk
 .PHONY: container
 container: GOOS = linux
 container: manager
-	docker build --load -t ${IMG} .
+	docker build --build-arg TARGETARCH=amd64 --load -t ${IMG} .
```
Comment (swiatekm) [inline Makefile:381]: Why are you changing these targets? Docker should build for the host architecture by default.

File: apis/v1alpha1/opentelemetrycollector_types.go [hunk 2]
```diff
@@ -390,15 +390,15 @@ type OpenTelemetryTargetAllocatorPrometheusCR struct {
...
+	// label selector matches no objects.
 	// +optional
 	PodMonitorSelector map[string]string `json:"podMonitorSelector,omitempty"`
 	// ServiceMonitors to be selected for target discovery.
-	// This is a map of {key,value} pairs. Each {key,value} in the map is going to exactly match a label in a
-	// ServiceMonitor's meta labels. The requirements are ANDed.
-	// Empty or nil map matches all service monitors.
+	// A label selector is a label query over a set of resources. The result of matchLabels and
+	// matchExpressions are ANDed. An empty label selector matches all objects. A null
+	// label selector matches no objects.
```
Comment (swiatekm) [inline apis/v1alpha1/opentelemetrycollector_types.go:401]: This is incorrect. In v1alpha1, these are in fact just maps, and the documentation is right. We switched to standard selectors in v1beta1, as this was a breaking change.

File: apis/v1beta1/targetallocator_types.go [hunk 3]
```diff
@@ -42,6 +42,18 @@ type TargetAllocatorPrometheusCR struct {
 	// label selector matches no objects.
 	// +optional
 	ServiceMonitorSelector *metav1.LabelSelector `json:"serviceMonitorSelector,omitempty"`
+	// ScrapeConfig to be selected for target discovery.
+	// A label selector is a label query over a set of resources. The result of matchLabels and
+	// matchExpressions are ANDed. An empty label selector matches all objects. A null
+	// label selector matches no objects.
+	// +optional
+	ScrapeConfigSelector *metav1.LabelSelector `json:"scrapeConfigSelector,omitempty"`
+	// ProbeSelector to be selected for target discovery.
```
Comment (swiatekm) [inline apis/v1beta1/targetallocator_types.go:51]:
```suggestion
	// Probes to be selected for target discovery.
```

File: apis/v1beta1/targetallocator_types.go [hunk 4]
```diff
@@ -42,6 +42,18 @@ type TargetAllocatorPrometheusCR struct {
 	// label selector matches no objects.
 	// +optional
 	ServiceMonitorSelector *metav1.LabelSelector `json:"serviceMonitorSelector,omitempty"`
+	// ScrapeConfig to be selected for target discovery.
```
Comment (swiatekm) [inline apis/v1beta1/targetallocator_types.go:45]:
```suggestion
	// ScrapeConfigs to be selected for target discovery.
```

**Review by swiatekm — COMMENTED** [review]
Looks good overall, some relatively minor nitpicks aside. Could you also add an e2e test for this behaviour? You can probably modify https://github.com/open-telemetry/opentelemetry-operator/tree/main/tests/e2e-targetallocator/targetallocator-prometheuscr.

File: Makefile (hunk 1 above)
Comment (dominicqi) [reply Makefile:381]: sorry  i test my local env, i will revert this

File: apis/v1beta1/targetallocator_types.go (hunk 3 above)
Comment (dominicqi) [reply apis/v1beta1/targetallocator_types.go]: fixed

**Comment by dominicqi** [comment]
@swiatekm e2e test added

**Comment by swiatekm** [comment]
> @swiatekm e2e test added

Thanks for adding it! What you've added checks that setting the attribute on the CR leads to a change in the target allocator ConfgMap, which is a part of it. But since this PR effectively exposes this functionality to users, I'd like the new e2e test to also check that we actually allocate targets derived from ScrapeConfigs and Probes. 

What I'd like you to add is the following:

* Create a ScrapeConfig and/or Probe targetting the Otel collector you create in the test
* Set the selector to select that ScrapeConfig/Probe
* Use Jobs similar to the [existing test](https://github.com/open-telemetry/opentelemetry-operator/blob/main/tests/e2e-targetallocator/targetallocator-prometheuscr/01-install.yaml) to verify that the target allocator is actually allocating targets generated for the ScrapeConfig/Probe

Does that make sense?

For the record, the rest of your PR looks good to me.

**Comment by jaronoff97** [comment]
@dominicqi thanks for adding this :D I was planning on doing this when I got back from PTO but glad to see this is already here 🥳 With the changes that Mikolaj mentioned above this should be g2g.

**Comment by jaronoff97** [comment]
looks like some test failures still, if you want to pair and figure them out or need a hand let me know!

**Comment by dominicqi** [comment]
> looks like some test failures still, if you want to pair and figure them out or need a hand let me know!

@jaronoff97 I'm a bit busy lately, it would be great if you could help ，thanks

**Comment by robertcoltheart** [comment]
Is this change needed to fully support `ScrapeConfig` resources? I was trying to figure out why my TA wasn't picking up the ScrapConfig resource with `v0.114.0`. Happy to contribute if so.

**Comment by jaronoff97** [comment]
@robertcoltheart yes, that's the case. I've had some limited availibility... @dominicqi do you want me to push up some changes via a PR to your branch or do you want to pair?

**Comment by dominicqi** [comment]
> @robertcoltheart yes, that's the case. I've had some limited availibility... @dominicqi do you want me to push up some changes via a PR to your branch or do you want to pair?

I have some time now, I will try to fix it

**Comment by dominicqi** [comment]
all e2e passed

File: tests/e2e-targetallocator/targetallocator-prometheuscr/02-install.yaml [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+      restartPolicy: OnFailure
+      containers:
+        - name: check-metrics
+          image: curlimages/curl
+          args:
+            - /bin/sh
+            - -c
+            - |
+              for i in $(seq 30); do
+                if curl -m 1 -s http://prometheus-cr-v1beta1-collector:9090/metrics | grep "otelcol"; then exit 0; fi
```
Comment (swiatekm) [inline tests/e2e-targetallocator/targetallocator-prometheuscr/02-install.yaml:75]: This doesn't look like it checks a metric scraped from either the Probe or ScrapeConfig you added.

Comment (dominicqi) [reply tests/e2e-targetallocator/targetallocator-prometheuscr/02-install.yaml:75]: yes, I will remove it

Comment (dominicqi) [reply tests/e2e-targetallocator/targetallocator-prometheuscr/02-install.yaml]: removed

**Review by swiatekm — APPROVED** [review]

**Comment by swiatekm** [comment]
@jaronoff97 could you also have a look?

**Review by jaronoff97 — APPROVED** [review]
Perfect!! Thank you so much @dominicqi :bow:

**Pull request merged**